doc/en/fixture.rst

```diff
@@ -149,442 +627,966 @@ marked ``smtp_connection`` fixture function.  Running the test looks like this:
+we wouldn't be able to know if ``order.extend([2])`` or ``order += [3]`` would
+also have problems. After ``append_first`` throws an exception, pytest won't run
+any more fixtures for ``test_order``, and it won't even try to run
+``test_order`` itself. The only things that would've run would be ``order`` and
+``append_first``.
+
+
+
+
+.. _`finalization`:
+
+Teardown/Cleanup (AKA Fixture finalization)
+-------------------------------------------
 
-    test_smtpsimple.py F                                                 [100%]
+When we run our tests, we'll want to make sure they clean up after themselves so
+they don't mess with any other tests (and also so that we don't leave behind a
+mountain of test data to bloat the system). Fixtures in pytest offer a very
+useful teardown system, which allows us to define the specific steps necessary
+for each fixture to clean up after itself.
+
+This system can be leveraged in two ways.
+
+.. _`yield fixtures`:
+
+1. ``yield`` fixtures (recommended)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+"Yield" fixtures are those that ``yield`` instead of ``return``. With these
+fixtures, we can run some code and pass an object back to the requesting
+fixture/test, just like with the other fixtures. The only differences are:
+
+1. ``return`` is swapped out for ``yield``.
+2. Any teardown code for that fixture is placed *after* the ``yield``.
+
+pytest does its best to put all the fixtures for a given test in a linear order
+so that it can see which fixture happens first, second, third, and so on. Once
+this is done, pytest will run each fixture up until it returns or yields, and
+then move on to the next fixture in the list to do the same thing.
+
+Once the test is finished, pytest will go back down the list of fixtures, but in
+the *reverse order*, taking each one that yielded, and running the code inside
+it that was *after* the ``yield`` statement.
+
+As a simple example, let's say we want to test sending email from one user to
+another. We'll have to first make each user, then send the email from one user
+to the other, and finally assert that the other user received that message in
+their inbox. If we want to clean up after the test runs, we'll likely have to
+make sure the other user's mailbox is emptied before deleting that user,
+otherwise the system may complain.
+
+Here's what that might look like:
+
+.. code-block:: python
+
+    import pytest
+
+    from emaillib import Email, MailAdminClient
+
+
+    @pytest.fixture
+    def mail_admin():
+        return MailAdminClient()
+
+
+    @pytest.fixture
+    def sending_user(mail_admin):
+        user = mail_admin.create_user()
+        yield user
+        admin_client.delete_user(user)
+
+    @pytest.fixture
+    def receiving_user(mail_admin):
+        user = mail_admin.create_user()
+        yield user
+        admin_client.delete_user(user)
+
+    def test_email_received(receiving_user, email):
```

Comment:
I don't see an `email` fixture defined. And `sending_user` is used without being requested.

Reply:
Btw: if we see other examples in the docs, we can see `# content of XXX.py`: that's not only informative, but is actually used by [regendoc](https://github.com/pytest-dev/regendoc/) to run that example. Also blocks that start with `$` are considered shell commands, and the output is appended to the text below it.

One can use `tox -e regen` locally to get that output updated.